stdlib/LinearAlgebra/src/LinearAlgebra.jl

```diff
@@ -582,7 +582,11 @@ function __init__()
     Base.at_disable_library_threading(() -> BLAS.set_num_threads(1))
 
     if !haskey(ENV, "OPENBLAS_NUM_THREADS")
-        BLAS.set_num_threads(max(1, Sys.CPU_THREADS ÷ 2))
+        if Sys.ARCH === :aarch64
```

Comment:
> None of the common ARM CPUs that come to mind have hyperthreading.

Probably it's only/mainly Intel which does hyperthreading no?  You're excluding 32-bit ARM, which I also doubt does hyperthreading

Reply:
AMD Cpus also have it (it's called SMT for them). Power I think does it too. For someone running asahi for example this is probably wrong, same for someone running the new Intel ones. But it was wrong before and i don't think we figured out a way to query performance cores on linux.

Reply:
> Power I think does it too.

Power does 4x or 8x, depending on the chip, meaning we may want to divide by 4 or 8 for Power? Is that detectable?

> For someone running asahi for example this is probably wrong

On asahi, would the 4big4L read as 8 threads, and 8big2L read as 10 threads?
Just by chance, 4 would've been correct for the former, but 5 for the latter isn't exactly ideal either. I'd be curious whether 5 or 10 are faster.

>  It's only on CPUs where we know `Sys.CPU_THREADS` is already the "right" number of threads for BLAS that we should use `Sys.CPU_THREADS` instead of `Sys.CPU_THREADS ÷ 2`

I think this is a rather strong statement, in that 5 doesn't seem obviously better than 10 for 8 big core + 2 little core systems.

> Also, this should use `@static`

Seems unnecessary when using `Sys.ARCH`, given the syntax is valid and the compiler can optimize it away just fine.
```julia
julia> foo() = Sys.ARCH === :aarch64
foo (generic function with 1 method)

julia> @code_typed foo()
CodeInfo(
1 ─     return false
) => Bool
```
but I can use `@static` when switching to `Base.BinaryPlatforms.arch(Base.BinaryPlatforms.HostPlatform()) == "aarch64"`

> I think a better test is `Sys.isapple()`

Do you think Asahi linux is more common than Raspberry Pi 4, Graviton, Ampere, A64FX?

Not sure about Nvidea's products.
If we start seeing more ARM laptops, they'll probably have performance + efficiency cores for Windows and Linux.

So it's possible, but I do not know the market shares here.


There are Intel CPUs without hyperthreading, and you can also turn it off in the bios for CPUs with it (this is reasonably common in HPC environments).
So combined with the P+E cores from Alder Lake and beyond, this all does get pretty messy. Would be interesting to see benchmarks there.
Intel advertises their E cores as being efficient in terms of performance/area, not performance/power, so they're marketing their E cores as improving multithreading performance. Their next generation of CPUs will add a lot more E cores without adding extra P cores. Maybe we really do want to use 1 BLAS thread per E core, hoping that the BLAS library does reasonable load balancing (which isn't a given; MKL seems to, but I wouldn't bet on OpenBLAS or BLIS). We'd still probably only want 1 thread per P core, and these unfortunately do have hyperthreading...

Reply:
Alder lake is really annoying because the E cores don't have SMT but the P cores do. Ideally we should always find the number of P cores and set NUM_THREADS to that.

A quick test shows that, at least for BLAS, adding the E cores makes it quite a bit slower.
```julia 
versioninfo()
Julia Version 1.9.0-DEV.688
Commit 3eaed8b54c (2022-05-31 15:31 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 16 × 12th Gen Intel(R) Core(TM) i5-12600K
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, alderlake)
  Threads: 6 on 16 virtual cores
Environment:
  JULIA_NUM_THREADS = 6
  JULIA_NUM_PRECOMPILE_TASKS = 6

julia> using LinearAlgebra
A1=rand(8192,8192); A2=copy(A1);

julia> @btime lu!($A1);
  1.022 s (2 allocations: 64.05 KiB)

julia> BLAS.set_num_threads(10)

julia> @btime lu!($A2);
  1.847 s (2 allocations: 64.05 KiB)

```

Reply:
What about with `using MKL`? I'm curious if MKL does better.

Reply:
I see you also set `JULIA_NUM_PRECOMPILE_TASKS = 6`, do you think it even hurts this?

Reply:
> So mkl might do something not to use them

That seems likely.